File: August-LeetCoding-Challenge/09-Rotting-Oranges/Rotting-Oranges.cs
public class Solution {
    public int OrangesRotting(int[][] grid) {
        if(grid == null || grid.Length == 0) return 0;
        var rows = grid.Length;
        var cols = grid[0].Length;
        var count_fresh = 0;
        Queue<int[]> queue = new Queue<int[]>();
        //Push rotten oranges in the queue, count fresh oranges
        for(int i=0; i<rows; i++){
            for(int j=0; j<cols; j++){
                if(grid[i][j] ==2){
                    queue.Enqueue(new int[2]{i, j});
                }else if(grid[i][j] ==1){
                    count_fresh++;
                }
            }
        }
        
        if(count_fresh == 0) return 0;
        
        var count = 0;
        var dirs = new int[4][]{new int[]{1,0}, new int[]{-1,0}, new int[]{0,-1}, new int[]{0,1}};
        while(queue.Count > 0){
            count++;
            var size = queue.Count;
            for(int i=0; i<size; i++){
                var point = queue.Dequeue();
                foreach(int[] dir in dirs){
                    var x = point[0] + dir[0];
                    var y = point[1] + dir[1];
                    if(x<0 || y<0 || x>=rows || y>=cols || grid[x][y] == 0 || grid[x][y] == 2) continue;
                    
                    grid[x][y] =2;
                    queue.Enqueue(new int[2]{x, y});
                    count_fresh--;
                }
            }
        }
        return count_fresh == 0 ? count-1 : -1;
    }
}

File: August-LeetCoding-Challenge/09-Rotting-Oranges/Rotting-Oranges.kt
package breadth_first_search

import java.util.*

class RottingOrangesKotlin994 {
    fun orangesRotting(grid: Array<IntArray>): Int {
        val visited = Array(grid.size) { BooleanArray(grid[0].size) }
        val queue: Queue<IntArray> = LinkedList()
        // add all rotten orange into the queue,visited
        var freshOranges = 0
        for (i in grid.indices) {
            for (j in grid[0].indices) {
                if (grid[i][j] == 2) {
                    queue.offer(intArrayOf(i, j))
                    visited[i][j] = true
                }
                if (grid[i][j] == 1) {
                    ++freshOranges
                }
            }
        }
        if (freshOranges == 0) {
            return 0
        }
        // bfs
        var level = -1
        var count = 0
        while (queue.isNotEmpty()) {
            for (size in queue.indices) {
                val current = queue.poll()
                for (index in 0..3) {
                    val nextX = current[0] + deltaX[index]
                    val nextY = current[1] + deltaY[index]
                    if (inBound(grid, nextX, nextY) &&
                        !visited[nextX][nextY] &&
                        grid[nextX][nextY] == 1
                    ) {
                        queue.offer(intArrayOf(nextX, nextY))
                        visited[nextX][nextY] = true
                        ++count
                    }
                }
            }
            ++level
        }
        return if (count == freshOranges) level else -1
    }

    private val deltaX = intArrayOf(0, 0, -1, 1)
    private val deltaY = intArrayOf(-1, 1, 0, 0)
    private fun inBound(
        grid: Array<IntArray>,
        x: Int,
        y: Int
    ) = x >= 0 && y >= 0 && x < grid.size && y < grid[0].size
}

fun main() {
    val solution = RottingOrangesKotlin994()
    // 4
    println(
        solution.orangesRotting(
            arrayOf(
                intArrayOf(2, 1, 1),
                intArrayOf(1, 1, 0),
                intArrayOf(0, 1, 1)
            )
        )
    )
    // -1
    println(
        solution.orangesRotting(
            arrayOf(
                intArrayOf(2, 1, 1),
                intArrayOf(0, 1, 1),
                intArrayOf(1, 0, 1)
            )
        )
    )
}